### PR TITLE
Use framework prefilter for massive perf improvement when running fra…

### PR DESCRIPTION
**Pull request by MatthewBeardmore**
Related to #529

Implements creation of a prefilter only when VS is passing in a list of test cases. As OsirisTerje is already working on prefiltering with the TfsTestFilter, I didn't work on implementing anything else with preflighting in that method. VS seems to pass in a list of test cases when you are running a single test, which does not pass in a TfsTestFilter, and therefore, shouldn't affect his work.

### CODE REVIEW

**Comment by jnm2** [comment]
@OsirisTerje how to you think this should be tested?

File: src/NUnitTestAdapter/NUnitTestAdapter.cs [hunk 1]
```diff
@@ -166,6 +168,21 @@ private TestPackage CreateTestPackage(string assemblyName)
...
+            {
+                var prefilters = new List<string>();
+                
+                foreach (TestCase testCase in testCases)
+                {
+                    int end = testCase.FullyQualifiedName.IndexOfAny(new char[] { '(', '<' });
+                    if (end > 0)
+                        prefilters.Add(testCase.FullyQualifiedName.Substring(0, end));
+                    else
+                        prefilters.Add(testCase.FullyQualifiedName);
```
[the remark this answers is not in the document]
Comment (MatthewBeardmore) [reply src/NUnitTestAdapter/NUnitTestAdapter.cs]: I took this code from the NUnitLite code as I was looking for how it did prefiltering. Based on my (very limited) understanding of how VS is working here, it should always match what NUnit generates? But I'm not 100% sure either.

**Comment by OsirisTerje** [comment]
@MatthewBeardmore  Can you fix the conflicts ?  

Would like to get this into the 3.13 release this month.

**Comment by Dreamescaper** [comment]
@OsirisTerje 
Is there anything apart from conflicts preventing from merging?
I'd like it to be merged, so willing to pick it up (unless @MatthewBeardmore wants to finish it, of course).

**Comment by OsirisTerje** [comment]
@Dreamescaper Very good if you can continue on this.  You will have to make your own PR though.  
The conflicts are one thing, should be easy to fix though.
But  it really needs proper testing, so if you can get something done there to ensure we don't break anything with this, that would be awesome.
It should be tested both in VS and  on command line (using vstest.console and dotnet test) after the  changes.

**Comment by Dreamescaper** [comment]
I'll take a closer look, and will investigate if existing filtering coverage should be extended.

I'll clarify, though.
> It should be tested both in VS and on command line (using vstest.console and dotnet test) after the changes.

Are we talking about some manual checklist, or covered as integration tests?

**Comment by OsirisTerje** [comment]
We have some acceptance tests, see cake acceptance tests script and the separate project for that, but it doesn't cover all cases, so yes, we still have to do manual tests. 
The other unit tests may cover part of this, but I haven't checked in this case.

**Comment by MatthewBeardmore** [comment]
Sorry, I've been busy for months... I've fixed up the merge conflicts. Let me know if there's anything else that you need. I realize this isn't comprehensive testing by any means, but I've been using this change internally since I created this branch and haven't run into any issues at this point.

**Comment by OsirisTerje** [comment]
@MatthewBeardmore The prerelease is now available at Install-Package NUnit3TestAdapter -Version 3.15.0-dev-01118 -Source https://www.myget.org/F/nunit/api/v3/index.json  .  Can you check this out?

I'm planning on releasing this separately next week, but will like some more checks on it.  This issue and PR will then be the only one in that release.

**Comment by MatthewBeardmore** [comment]
Sure, I'm adding it today.

**Comment by OsirisTerje** [comment]
@MatthewBeardmore  Any news ?

**Comment by MatthewBeardmore** [comment]
I haven't seen any issues with it so far.

**Comment by OsirisTerje** [comment]
Cool, thanks!  I'll get it released within a few days.

**Pull request merged**